subprojects/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/DefaultPropertyAnnotationMetadata.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.reflect.annotations.impl;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.gradle.internal.reflect.ParameterValidationContext;
+import org.gradle.internal.reflect.annotations.PropertyAnnotationCategory;
+import org.gradle.internal.reflect.annotations.PropertyAnnotationMetadata;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nullable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+public class DefaultPropertyAnnotationMetadata implements PropertyAnnotationMetadata {
+    private final String propertyName;
+    private final Method getter;
+    private final ImmutableMap<PropertyAnnotationCategory, Annotation> annotations;
+    private final ImmutableList<String> validationProblems;
+
+    public DefaultPropertyAnnotationMetadata(String propertyName, Method getter, ImmutableMap<PropertyAnnotationCategory, Annotation> annotations, ImmutableList<String> validationProblems) {
+        this.propertyName = propertyName;
+        this.getter = getter;
+        this.annotations = annotations;
+        this.validationProblems = validationProblems;
+    }
+
+    @Override
+    public String getPropertyName() {
+        return propertyName;
+    }
+
+    @Override
+    public Method getGetter() {
+        return getter;
+    }
+
+    @Nullable
+    @Override
+    public Annotation getAnnotation(PropertyAnnotationCategory category) {
+        return annotations.get(category);
+    }
+
+    @Override
+    public boolean hasAnnotation(PropertyAnnotationCategory category, Class<? extends Annotation> annotationType) {
+        Annotation annotation = annotations.get(category);
+        return annotation != null && annotation.annotationType().equals(annotationType);
+    }
+
+    @Override
+    public Map<PropertyAnnotationCategory, Annotation> getAnnotations() {
+        return annotations;
+    }
+
+    @Override
+    public void visitValidationFailures(@Nullable String ownerPath, ParameterValidationContext validationContext) {
+        validationProblems.forEach(validationProblem -> validationContext.visitError(ownerPath, propertyName, validationProblem));
```

Comment:
I'm really unsure about where to draw the line here. Do you have some good material to read about that JIT compilation comparison? Also, reverse iteration with indexes should be faster for some cases, yet we don't use it because the code is nicer with iterators.

Reply:
I only have some anecdotal evidence: https://stackoverflow.com/a/20177092

I didn't make up my mind on whether using a `for` loop or `foreach` is better, I am just wondering if there is a good rule of thumb which one to use.

Reply:
I think we should discuss this, but perhaps not on this PR. From my personal point of view they have different advantages/disadvantages. Lambdas are much more concise, but don't have a `break` construct, and `continue` looks weird, too, so you really have to use `filter()` and other tools as well. I think we can't really adopt lambdas properly until we can switch to Java 8-compatible Guava, as many use cases require something like `.collect(toImmutableList())`.